### ISSUE TITLE
Overriding descriptors without a `__set__` (in particular `@cached_property`)

### ISSUE THREAD

**Issue by thomkeh** [microsoft/pyright#4438]
**Is your feature request related to a problem? Please describe.**
If a descriptor doesn't provide a `__set__` method, it can be replaced by a simple assignment.

This is for example the case with the standard library's [cached_property](https://docs.python.org/3/library/functools.html#functools.cached_property). The documentation states:
> A regular `property` blocks attribute writes unless a setter is defined. In contrast, a `cached_property` allows writes.

And indeed this code works fine:
```python
from functools import cached_property

class B:
    @cached_property
    def x(self) -> int:
        return 1234 * 234

class C(B):
    def __init__(self):
        self.x = 2

b = B()
print(b.x)  # 288756
c = C()
print(c.x)  # 2
```
However, pyright complains:
```
error: Cannot assign member "x" for type "C"
    "Literal[2]" is incompatible with "cached_property[int]" (reportGeneralTypeIssues)
```

**Describe the solution you'd like**
If no `__set__` is defined, then pyright could perhaps look at the return type of `__get__` and allow assigning values of that type?

There might be other problems with this that I'm not foreseeing right now.

**Comment by erictraut** [comment]
The current behavior is correct from a type checking perspective. The type `B.x` has a declared type, and you are attempting to assign it an incompatible type within a subclass. This doesn't generate an exception at runtime, but it is a type violation, hence the error.
